.github/workflows/build.yml

```diff
@@ -484,6 +484,31 @@ jobs:
     - name: Tests
       run: xvfb-run make test
 
+  build_tsan_free_threading:
+    name: 'Thread sanitizer (free-threading)'
```

Comment:
Can we have another TSan build with the GIL enabled? We can still have low-level synchronization bugs in the codebase.

Reply:
Okay, I will try to set up it also. (I am still struggling with GA :( )

Reply:
@pitrou 

FYI, at the GIL build following errors are found.
But I am going to disable test_logging from tsan test for temporary, because it looks like some investigation is needed for free-threaded build too (binding issue)

```
0:00:51 load avg: 3.02 [10/14/1] test_logging failed (1 error, 1 failure) -- running (1): test_io (30.8 sec)
==================
WARNING: ThreadSanitizer: thread leak (pid=17430)
  Thread T1 (tid=17410, finished) created by main thread at:
    #0 pthread_create ../../../../src/libsanitizer/tsan/tsan_interceptors_posix.cpp:969 (libtsan.so.0+0x605b8)
    #1 do_start_joinable_thread Python/thread_pthread.h:284 (python+0x40c430)
    #2 PyThread_start_joinable_thread Python/thread_pthread.h:308 (python+0x40c430)
    #3 do_start_new_thread Modules/_threadmodule.c:1430 (python+0x4c5580)
    #4 thread_PyThread_start_joinable_thread Modules/_threadmodule.c:1521 (python+0x4c5aeb)
    #5 _PyEval_EvalFrameDefault Python/generated_cases.c.h:1[18](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:19)1 (python+0x92eb5)
    #6 _PyEval_EvalFrame Include/internal/pycore_ceval.h:114 (python+0x33e7bf)
    #7 _PyEval_Vector Python/ceval.c:1820 (python+0x33e7bf)
    #8 _PyFunction_Vectorcall Objects/call.c:413 (python+0x13d32b)
    #9 _PyObject_VectorcallTstate Include/internal/pycore_call.h:168 (python+0x1440c0)
    #10 method_vectorcall Objects/classobject.c:92 (python+0x1440c0)
    #11 _PyVectorcall_Call Objects/call.c:273 (python+0x140820)
    #12 _PyObject_Call Objects/call.c:348 (python+0x140820)
    #13 PyObject_Call Objects/call.c:373 (python+0x140820)
    #14 _PyEval_EvalFrameDefault Python/generated_cases.c.h:1270 (python+0x83e4e)
    #15 _PyEval_EvalFrame Include/internal/pycore_ceval.h:114 (python+0x33e7bf)
    #16 _PyEval_Vector Python/ceval.c:1820 (python+0x33e7bf)
    #17 _PyFunction_Vectorcall Objects/call.c:413 (python+0x13d32b)
    #18 _PyObject_VectorcallDictTstate Objects/call.c:135 (python+0x14159c)
    #[19](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:20) _PyObject_Call_Prepend Objects/call.c:504 (python+0x14159c)
    #[20](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:21) slot_tp_call Objects/typeobject.c:9176 (python+0x24e1da)
    #[21](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:22) _PyObject_MakeTpCall Objects/call.c:242 (python+0x13c186)
    #[22](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:23) _PyObject_VectorcallTstate Include/internal/pycore_call.h:166 (python+0x13d038)
    #[23](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:24) _PyObject_VectorcallTstate Include/internal/pycore_call.h:153 (python+0x13d038)
    #[24](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:25) PyObject_Vectorcall Objects/call.c:327 (python+0x13d038)
    #[25](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:26) _PyEval_EvalFrameDefault Python/generated_cases.c.h:817 (python+0x9174f)
    #[26](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:27) _PyEval_EvalFrame Include/internal/pycore_ceval.h:114 (python+0x33e7bf)
    #[27](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:28) _PyEval_Vector Python/ceval.c:1820 (python+0x33e7bf)
    #[28](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:29) _PyFunction_Vectorcall Objects/call.c:413 (python+0x13d32b)
    #[29](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:30) _PyObject_VectorcallTstate Include/internal/pycore_call.h:168 (python+0x1440c0)
    #[30](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:31) method_vectorcall Objects/classobject.c:92 (python+0x1440c0)
    #[31](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:32) _PyVectorcall_Call Objects/call.c:273 (python+0x140820)
    #[32](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:33) _PyObject_Call Objects/call.c:348 (python+0x140820)
    #[33](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:34) PyObject_Call Objects/call.c:373 (python+0x140820)
    #[34](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:35) _PyEval_EvalFrameDefault Python/generated_cases.c.h:1270 (python+0x83e4e)
    #[35](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:36) _PyEval_EvalFrame Include/internal/pycore_ceval.h:114 (python+0x33e7bf)
    #[36](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:37) _PyEval_Vector Python/ceval.c:1820 (python+0x33e7bf)
    #[37](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:38) _PyFunction_Vectorcall Objects/call.c:413 (python+0x13d32b)
    #[38](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:39) _PyObject_VectorcallDictTstate Objects/call.c:135 (python+0x14159c)
    #[39](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:40) _PyObject_Call_Prepend Objects/call.c:504 (python+0x14159c)
    #[40](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:41) slot_tp_call Objects/typeobject.c:9176 (python+0x24e1da)
    #[41](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:42) _PyObject_MakeTpCall Objects/call.c:2[42](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:43) (python+0x13c186)
    #42 _PyObject_VectorcallTstate Include/internal/pycore_call.h:166 (python+0x13d038)
    #[43](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:44) _PyObject_VectorcallTstate Include/internal/pycore_call.h:153 (python+0x13d038)
    #[44](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:45) PyObject_Vectorcall Objects/call.c:327 (python+0x13d038)
    #[45](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:46) _PyEval_EvalFrameDefault Python/generated_cases.c.h:817 (python+0x9174f)
    #[46](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:47) _PyEval_EvalFrame Include/internal/pycore_ceval.h:114 (python+0x33e7bf)
    #[47](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:48) _PyEval_Vector Python/ceval.c:1820 (python+0x33e7bf)
    #[48](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:49) _PyFunction_Vectorcall Objects/call.c:413 (python+0x13d32b)
    #[49](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:50) _PyObject_VectorcallTstate Include/internal/pycore_call.h:168 (python+0x1440c0)
    #[50](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:51) method_vectorcall Objects/classobject.c:92 (python+0x1440c0)
    #[51](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:52) _PyVectorcall_Call Objects/call.c:273 (python+0x140820)
    #[52](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:53) _PyObject_Call Objects/call.c:348 (python+0x140820)
    #[53](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:54) PyObject_Call Objects/call.c:373 (python+0x140820)
    #[54](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:55) _PyEval_EvalFrameDefault Python/generated_cases.c.h:1270 (python+0x83e4e)
    #[55](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:56) _PyEval_EvalFrame Include/internal/pycore_ceval.h:114 (python+0x33e7bf)
    #[56](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:57) _PyEval_Vector Python/ceval.c:1820 (python+0x33e7bf)
    #[57](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:58) _PyFunction_Vectorcall Objects/call.c:413 (python+0x13d32b)
    #[58](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:59) _PyObject_VectorcallDictTstate Objects/call.c:135 (python+0x141[59](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:60)c)
    #59 _PyObject_Call_Prepend Objects/call.c:504 (python+0x14159c)
    #[60](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:61) slot_tp_call Objects/typeobject.c:9176 (python+0x24e1da)
    #[61](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:62) _PyObject_MakeTpCall Objects/call.c:242 (python+0x13c186)
    #[62](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:63) _PyObject_VectorcallTstate Include/internal/pycore_call.h:166 (python+0x13d038)
    #[63](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:64) _PyObject_VectorcallTstate Include/internal/pycore_call.h:153 (python+0x13d038)
    #[64](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:65) PyObject_Vectorcall Objects/call.c:327 (python+0x13d038)
    #[65](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:66) _PyEval_EvalFrameDefault Python/generated_cases.c.h:817 (python+0x9174f)
    #[66](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:67) _PyEval_EvalFrame Include/internal/pycore_ceval.h:114 (python+0x33e4e0)
    #[67](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:68) _PyEval_Vector Python/ceval.c:1820 (python+0x33e4e0)
    #[68](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:69) PyEval_EvalCode Python/ceval.c:599 (python+0x33e4e0)
    #[69](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:70) builtin_exec_impl Python/bltinmodule.c:1132 (python+0x3371a8)
    #[70](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:71) builtin_exec Python/clinic/bltinmodule.c.h:521 (python+0x33[71](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:72)a8)
    #71 cfunction_vectorcall_FASTCALL_KEYWORDS Objects/methodobject.c:441 (python+0x1da196)
    #[72](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:73) _PyObject_VectorcallTstate Include/internal/pycore_call.h:168 (python+0x13cf8a)
    #[73](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:74) PyObject_Vectorcall Objects/call.c:327 (python+0x13cf8a)
    #[74](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:75) _PyEval_EvalFrameDefault Python/generated_cases.c.h:817 (python+0x9174f)
    #[75](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:76) _PyEval_EvalFrame Include/internal/pycore_ceval.h:114 (python+0x33e7bf)
    #[76](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:77) _PyEval_Vector Python/ceval.c:1820 (python+0x33e7bf)
    #[77](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:78) _PyFunction_Vectorcall Objects/call.c:413 (python+0x13d32b)
    #[78](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:79) _PyVectorcall_Call Objects/call.c:273 (python+0x140820)
    #[79](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:80) _PyObject_Call Objects/call.c:348 (python+0x140820)
    #[80](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:81) PyObject_Call Objects/call.c:373 (python+0x140820)
    #[81](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:82) pymain_run_module Modules/main.c:297 (python+0x4297e5)
    #[82](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:83) pymain_run_python Modules/main.c:622 (python+0x42a3ec)
    #[83](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:84) Py_RunMain Modules/main.c:707 (python+0x42b1c5)
    #[84](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:85) pymain_main Modules/main.c:737 (python+0x42b1c5)
    #[85](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:86) Py_BytesMain Modules/main.c:761 (python+0x42b1c5)
    #[86](https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872#step:9:87) main Programs/python.c:15 (python+0x81dc5)

SUMMARY: ThreadSanitizer: thread leak Python/thread_pthread.h:284 in do_start_joinable_thread
```

Reply:
see: https://github.com/python/cpython/actions/runs/8300251375/job/22717784893?pr=116872